### PR TITLE
Repetitions Model

### PR DESCRIPTION
**Pull request by AntonioFernandezViz**
Gymnos components of repetition model based on utterances.

Curently, this model used Gymnos components to search the optimal supervised model that fits a local dataset loaded outside Gymnos (in next steps a dataset class will be deploy) with binary labels and sequences of phrases as features.

   - preprocessors:
        "utterances_aura_embeddings": Apply a simple preprocessing to the sequences and an embedding trained in a .pkl file from Aura logs.
        "utterances_embedding_pooling": Create a Pooling from arrays based on different types of concatenations of statistics.
   - models: Several classifiers of sklearn used in the repetition model and grids file.
   - Registration of preprocessors and models.

### CODE REVIEW

File: gymnos/models/repetition_ada_boost.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        Number of chunks in cross validation
+    search: str
+        Type of hyperparameters search (grid search or random search)
+
+    Note
+    ----
+    This model requires binary labels.
+    """
+
+    def __init__(self, cv, search):
```
Comment (r-salas) [inline gymnos/models/repetition_ada_boost.py]: Add default parameters, for example `cv=5` and `search=None`

File: gymnos/models/repetition_ada_boost.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+            model_search = RandomizedSearchCV(estimator=model_search, param_distributions=ADA_BOOST_RANDOM_GRID,
+                                              scoring='roc_auc', cv=self.cv, refit=True,
+                                              random_state=314, verbose=3)
+        else:
+            pass
+        model_search.fit(X, y)
+        if self.search in ["grid_search", "random_search"]:
+            self.model = model_search.best_estimator_
+
+    def fit_generator(self, generator):
```
Comment (r-salas) [inline gymnos/models/repetition_ada_boost.py]: Your model doesn't support generators, you should not implement this method. By default an error is raised if you're trying to call `fit_generator` with a model that does not support generators

File: gymnos/models/repetition_ada_boost.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        return {}
+
+    def predict(self, X):
+        return self.model.predict(X)
+
+    def evaluate(self, X, y):
+        result = self.predict(X)
+        cr = classification_report(y, result, output_dict=True)
+        probs = self.model.predict_proba(X)[:, 1]
```
Comment (r-salas) [inline gymnos/models/repetition_ada_boost.py]: Implement the `predict_proba` method. Don't return probabilities in this method. This method should only return metrics .

File: gymnos/models/repetition_light_gbm.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
+#
+#
+#   Repetition LightGBM
+#
+#
+
+import lightgbm as lgb
```
Comment (r-salas) [inline gymnos/models/repetition_light_gbm.py]: This module is not installed with Gymnos. Read documentation for lazy imports (Guide/Lazy Imports)

File: gymnos/models/repetition_light_gbm.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+from sklearn.metrics import classification_report, roc_auc_score
+from sklearn.model_selection import GridSearchCV, RandomizedSearchCV
+
+from .mixins import SklearnMixin
+from .model import Model
+from .utils.repetition_grids import LIGHT_GBM_RANDOM_GRID, LIGHT_GBM_GRID
+
+
+class RepetitionLightGBM(SklearnMixin, Model):
```
Comment (r-salas) [inline gymnos/models/repetition_light_gbm.py:15]: You're overriding all the methods `SklearnMixin` implements for you. You don't need to inherit from SklearnMixin. Please read documentation for mixins (Guide/Add a model#Scikit-Learn mixin)

File: gymnos/models/repetition_xgboost.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
+#
+#
+#   Repetition AdaBoost
+#
+#
+
+import xgboost as xgb
```
Comment (r-salas) [inline gymnos/models/repetition_xgboost.py]: Dependency should be a lazy import

File: gymnos/models/utils/repetition_grids.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
+import numpy as np
```
Comment (r-salas) [inline gymnos/models/utils/repetition_grids.py]: This file looks like constants for specific models, maybe move each constant to each model

File: gymnos/models/repetition_knn.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
+#
```
Comment (r-salas) [inline gymnos/models/repetition_knn.py:1]: Same as `repetition_ada_boost`

File: gymnos/models/repetition_random_forest.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
+#
```
Comment (r-salas) [inline gymnos/models/repetition_random_forest.py:1]: Same as `repetition_ada_boost` and `light_gbm`

File: gymnos/models/repetition_svm.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
+#
```
Comment (r-salas) [inline gymnos/models/repetition_svm.py]: Same as `repetition_ada_boost` and `light_gbm`

File: gymnos/preprocessors/utterances_aura_embeddings.py [hunk 11]
```diff
@@ -0,0 +1,60 @@
+#
+#
+#   Utterances Aura Embeddings
+#
+#
+
+import string
+
+import unidecode
+from auracog_embeddings.embeddings import Embeddings
```
Comment (r-salas) [inline gymnos/preprocessors/utterances_aura_embeddings.py]: This looks like an Aura internal module. Add module to lazy imports. I guess this module is installable with a private PyPi server, please add documentation about how to add this private PyPi server

File: gymnos/preprocessors/utterances_embedding_pooling.py [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+                    total_phrase = [item for sublist in sequence for item in sublist]
+                else:
+                    total_phrase.append(
+                        UtterancesEmbeddingPooling.pooling(element=phrase, type_pooling=type_pooling_phrase))
+            total_sequence.append(
+                UtterancesEmbeddingPooling.pooling(element=total_phrase, type_pooling=type_pooling_sequence))
+        return total_sequence
+
+    @staticmethod
+    def pooling(element, type_pooling):
```
Comment (r-salas) [inline gymnos/preprocessors/utterances_embedding_pooling.py]: All models must have the same public API. Please write this method as a private method (e.g `_apply_pooling`). It doesn't look like you're using this method so if you're not using it, please remove this method.

File: gymnos/preprocessors/utterances_embedding_pooling.py [hunk 13]
```diff
@@ -0,0 +1,103 @@
+#
+#
+#   Utterances Embeddings Pooling
+#
+#
+
+import numpy as np
+
+from gymnos.preprocessors.preprocessor import Preprocessor
```
Comment (r-salas) [inline gymnos/preprocessors/utterances_embedding_pooling.py]: Please use relative imports

File: gymnos/preprocessors/utterances_aura_embeddings.py [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+
+from gymnos.preprocessors.preprocessor import Preprocessor
```
Comment (r-salas) [inline gymnos/preprocessors/utterances_aura_embeddings.py]: Please use relative imports

File: gymnos/preprocessors/utterances_embedding_pooling.py [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+    This class applies pooling to list of list of arrays concatenating several combinations of statistics.
+
+    Parameters
+    -----------
+    type_pooling: str,
+                Type of pooling (flatten_sequence, two_steps_complete_mean, two_steps_complete_mean_max
+                or two_steps_complete_complete)
+    """
+
+    def __init__(self, type_pooling):
```
Comment (r-salas) [inline gymnos/preprocessors/utterances_embedding_pooling.py]: Default value?

File: gymnos/preprocessors/utterances_embedding_pooling.py [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+        self.type_pooling = type_pooling
```
Comment (r-salas) [inline gymnos/preprocessors/utterances_embedding_pooling.py]: Assert value for `type_pooling` is valid

**Review by r-salas — CHANGES_REQUESTED** [review]

File: gymnos/preprocessors/utterances_embedding_pooling.py (hunk 16 above)
Comment (AntonioFernandezViz) [reply gymnos/preprocessors/utterances_embedding_pooling.py]: Done, thank you

File: gymnos/preprocessors/utterances_embedding_pooling.py (hunk 15 above)
Comment (AntonioFernandezViz) [reply gymnos/preprocessors/utterances_embedding_pooling.py]: Done, thank you

File: gymnos/preprocessors/utterances_aura_embeddings.py (hunk 14 above)
Comment (AntonioFernandezViz) [reply gymnos/preprocessors/utterances_aura_embeddings.py]: Done, thank you

File: gymnos/preprocessors/utterances_embedding_pooling.py (hunk 13 above)
Comment (AntonioFernandezViz) [reply gymnos/preprocessors/utterances_embedding_pooling.py]: Done, thank you

File: gymnos/preprocessors/utterances_embedding_pooling.py (hunk 12 above)
Comment (AntonioFernandezViz) [reply gymnos/preprocessors/utterances_embedding_pooling.py]: Done, thank you

File: gymnos/models/repetition_xgboost.py (hunk 6 above)
Comment (AntonioFernandezViz) [reply gymnos/models/repetition_xgboost.py]: Done, thank you

File: gymnos/models/utils/repetition_grids.py (hunk 7 above)
Comment (AntonioFernandezViz) [reply gymnos/models/utils/repetition_grids.py]: Done, thank you

File: gymnos/models/repetition_knn.py (hunk 8 above)
Comment (AntonioFernandezViz) [reply gymnos/models/repetition_knn.py:1]: Done, thank you

File: gymnos/models/repetition_random_forest.py (hunk 9 above)
Comment (AntonioFernandezViz) [reply gymnos/models/repetition_random_forest.py:1]: Done, thank you

File: gymnos/models/repetition_svm.py (hunk 10 above)
Comment (AntonioFernandezViz) [reply gymnos/models/repetition_svm.py]: Done, thank you

File: gymnos/preprocessors/utterances_aura_embeddings.py (hunk 11 above)
Comment (AntonioFernandezViz) [reply gymnos/preprocessors/utterances_aura_embeddings.py]: Done, thank you. In next pull request I will add documentation.

File: gymnos/models/repetition_light_gbm.py (hunk 5 above)
Comment (AntonioFernandezViz) [reply gymnos/models/repetition_light_gbm.py:15]: I need to inherit save and restore methods from Model so I keep SklearnMixIn as super class.

File: gymnos/models/repetition_light_gbm.py (hunk 4 above)
Comment (AntonioFernandezViz) [reply gymnos/models/repetition_light_gbm.py]: Done, thank you

File: gymnos/models/repetition_ada_boost.py (hunk 3 above)
Comment (AntonioFernandezViz) [reply gymnos/models/repetition_ada_boost.py]: Done, thank you

File: gymnos/models/repetition_ada_boost.py (hunk 2 above)
Comment (AntonioFernandezViz) [reply gymnos/models/repetition_ada_boost.py]: Done, thank you

File: gymnos/models/repetition_ada_boost.py (hunk 1 above)
Comment (AntonioFernandezViz) [reply gymnos/models/repetition_ada_boost.py]: Done, thank you

File: gymnos/preprocessors/utterances_embedding_pooling.py (hunk 16 above)
Comment (AntonioFernandezViz) [reply gymnos/preprocessors/utterances_embedding_pooling.py]: Done, thank you!

File: gymnos/preprocessors/utterances_embedding_pooling.py (hunk 15 above)
Comment (AntonioFernandezViz) [reply gymnos/preprocessors/utterances_embedding_pooling.py]: Done, thank you!

File: gymnos/preprocessors/utterances_aura_embeddings.py (hunk 14 above)
Comment (AntonioFernandezViz) [reply gymnos/preprocessors/utterances_aura_embeddings.py]: Done, thank you!

File: gymnos/preprocessors/utterances_embedding_pooling.py (hunk 13 above)
Comment (AntonioFernandezViz) [reply gymnos/preprocessors/utterances_embedding_pooling.py]: Done, thank you!

File: gymnos/preprocessors/utterances_embedding_pooling.py (hunk 12 above)
Comment (AntonioFernandezViz) [reply gymnos/preprocessors/utterances_embedding_pooling.py]: Done, thank you!

File: gymnos/models/repetition_light_gbm.py (hunk 5 above)
Comment (AntonioFernandezViz) [reply gymnos/models/repetition_light_gbm.py:15]: I need to inherit to use save and restore methods from SklearnMixIn class.

File: gymnos/models/repetition_light_gbm.py (hunk 4 above)
Comment (AntonioFernandezViz) [reply gymnos/models/repetition_light_gbm.py]: Done, thank you!

File: gymnos/models/repetition_ada_boost.py (hunk 3 above)
Comment (AntonioFernandezViz) [reply gymnos/models/repetition_ada_boost.py]: Done, thank you!

File: gymnos/models/repetition_ada_boost.py (hunk 2 above)
Comment (AntonioFernandezViz) [reply gymnos/models/repetition_ada_boost.py]: Done, thank you!

File: gymnos/models/repetition_ada_boost.py (hunk 1 above)
Comment (AntonioFernandezViz) [reply gymnos/models/repetition_ada_boost.py]: Done, thank you!

**Review by AntonioFernandezViz — COMMENTED** [review]
All changes have been done except:
- I do not erase SklearnMiin dependency because I need to inherit save and restore methods.
- Add documentation of import of auracog.embedding (in next pr I will update documentation).

File: gymnos/models/repetition_light_gbm.py [hunk 17]
```diff
@@ -29,40 +29,51 @@ class RepetitionLightGBM(SklearnMixin, Model):
     This model requires binary labels.
     """
 
-    def __init__(self, cv, search):
-        self.model = lgb.LGBMClassifier(max_depth=-1, random_state=314, silent=True, metric='None', n_jobs=4,
-                                        n_estimators=5000)
+    def __init__(self, cv=5, search=None):
+        self.model = lazy_imports.lightgbm.LGBMClassifier(max_depth=-1, random_state=314, silent=True, metric='None',
```
Comment (r-salas) [inline gymnos/models/repetition_light_gbm.py]: :ok_hand:

**Review by r-salas — DISMISSED** [review]
Great job :clap:! Documentation for lazy_imports.aura_cog and documentation for models/preprocessors is still pending. Please add docs in next PR.

File: gymnos/models/repetition_light_gbm.py (hunk 17 above)
Comment (AntonioFernandezViz) [reply gymnos/models/repetition_light_gbm.py]: Thanks!

**Review by kawaits — CHANGES_REQUESTED** [review]
Need further devel before getting merge. Apart form current review,
dataset and trainer integration is required.

Thanks!
